src/reader.rs

```diff
@@ -138,52 +95,163 @@ impl<'a, R: Read> Reader<'a, R> {
-                        for _ in 0..block_len {
-                            let item = from_avro_datum(
-                                &self.writer_schema,
-                                &mut block_bytes,
-                                self.reader_schema,
-                            )?;
-                            self.items.push_back(item)
-                        }
-
-                        return Ok(())
-                    }
+    fn read_block_next(&mut self) -> Result<(), Error> {
+        if !self.is_empty() {
+            // I suppose this isn't technically an error...
+            panic!("(TODO Replace this with an error) Invalid state; shouldn't call read_block_next");
+        }
+        match util::read_long(&mut self.reader) {
+            Ok(block_len) => {
+                self.message_count = block_len as usize;
+                let block_bytes = util::read_long(&mut self.reader)?;
+                self.fill_buf(block_bytes as usize)?;
+                let mut marker = [0u8; 16];
+                self.reader.read_exact(&mut marker)?;
+
+                if marker != self.marker {
+                    return Err(DecodeError::new(
+                        "block marker does not match header marker",
+                    ).into())
                 }
+
+                // TODO (JAB): This is bad, but will fix later.
+                // Can use some "limited read" type to decode directly into the buffer.
+                self.codec.decompress(&mut self.buf)?;
+                
+                return Ok(())
             },
             Err(e) => match e.downcast::<::std::io::Error>()?.kind() {
                 // to not return any error in case we only finished to read cleanly from the stream
                 ErrorKind::UnexpectedEof => return Ok(()),
+                // TODO (JAB): This variant should probably return an error?
                 _ => (),
             },
         };
         Err(DecodeError::new("unable to read block").into())
     }
+
+
+    pub fn len(&self) -> usize {
+        self.message_count
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn read_next(&mut self, read_schema: Option<&Schema>) -> Result<Option<Value>, Error> {
+        if self.is_empty() {
+            self.read_block_next()?;
+            if self.is_empty() {
+                return Ok(None);
+            }
+        }
+
+        let mut block_bytes = &self.buf[self.buf_idx..];
+        let b_original = block_bytes.len();
+        let item = from_avro_datum(
+            &self.writer_schema,
+            &mut block_bytes,
+            read_schema,
+        )?;
+        self.buf_idx += b_original - block_bytes.len();
+        self.message_count -= 1;
+        Ok(Some(item))
+    }
+}
+
+// NOTE: This Reader can only read Avro Object Containers, not an arbitrary stream of Avro
+// data. I'll need to implement that separately. Both can be implemented generically if
```

Comment:
Is this still a TODO?

To read an arbitrary stream of Avro data, one can just leverage `from_avro_datum` (which is publicly usable)

Reply:
Yes, you are correct. Removed the `NOTE`